File: ai-qa-automation-platform/apps/api/app/eval/metrics.py
"""DeepEval metric wiring — workspace-scoped, thresholds from settings.

Metrics (spec §7):
  - AnswerRelevancyMetric  (soft, 0.8)
  - FaithfulnessMetric     (hard, 0.9)   → groundedness
  - GEval Completeness     (0.75)
  - GEval Correctness      (hard on high-risk, 0.85)
  - Custom ToolSequenceAccuracy (0.9)   → all required tools called, no
                                        out-of-order destructive calls, no extras

Thresholds are read from workspace settings at run time and snapshotted into
the Run record — never hardcoded in the metric definitions.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

# DeepEval is optional at import time (heavy); import lazily inside functions.
try:
    from deepeval.metrics import AnswerRelevancyMetric, FaithfulnessMetric, GEval
    from deepeval.test_case import LLMTestCase

    _DEEPEVAL_AVAILABLE = True
except ImportError:  # pragma: no cover
    _DEEPEVAL_AVAILABLE = False

    class GEval:  # stub so ToolSequenceAccuracy can still be defined
        def __init__(self, *args, **kwargs):
            self.score = 0.0
            self.threshold = kwargs.get("threshold", 0.9)
            self.success = False

        def is_successful(self) -> bool:
            return self.success

        def measure(self, *args, **kwargs) -> None:
            pass

DEFAULT_THRESHOLDS: dict[str, float] = {
    "answer_relevancy": 0.8,
    "groundedness": 0.9,
    "completeness": 0.75,
    "correctness": 0.85,
    "tool_sequence_accuracy": 0.9,
}

# Metrics that block the release gate when they fail.
HARD_GATE_METRICS = {"groundedness", "correctness", "tool_sequence_accuracy"}


@dataclass
class EvalInput:
    """One LLMTestCase worth of data, assembled by the eval layer."""

    input: str  # user query / flow under test
    actual_output: str  # model or app response
    retrieval_context: list[str]  # groundedness context
    tools_called: list[str]  # actual tool sequence
    expected_tools: list[str]  # from Jira acceptance criteria (ground truth)
    expected_output: str | None = None  # golden answer (existing tests / Jira AC)
    high_risk: bool = False


@dataclass
class MetricResult:
    metric: str
    score: float
    threshold: float
    hard_gate: bool
    passed: bool


# Canonical snake_case names shared by thresholds, filters, and the dashboard.
# Class-name forms from DeepEval are normalized to these.
METRIC_ALIASES: dict[str, str] = {
    "answerrelevancymetric": "answer_relevancy",
    "answer_relevancy": "answer_relevancy",
    "faithfulnessmetric": "groundedness",
    "groundedness": "groundedness",
    "completeness": "completeness",
    "correctness": "correctness",
    "toolsequenceaccuracy": "tool_sequence_accuracy",
    "tool_sequence_accuracy": "tool_sequence_accuracy",
}


def canonical_metric_name(name: str) -> str:
    return METRIC_ALIASES.get(name, name)


def evaluate_case(case: EvalInput, thresholds: dict[str, float] | None = None) -> list[MetricResult]:
    """Run all five metrics against one case. Returns per-metric results."""
    if not _DEEPEVAL_AVAILABLE:
        # Offline/eval-stub mode — deterministic pass-through for tests.
        return _stub_results(case, thresholds)

    t = {**DEFAULT_THRESHOLDS, **(thresholds or {})}
    tc = LLMTestCase(
        input=case.input,
        actual_output=case.actual_output,
        retrieval_context=case.retrieval_context,
        tools_called=case.tools_called,
        expected_tools=case.expected_tools,
        expected_output=case.expected_output or case.actual_output,
    )

    answer_relevancy = AnswerRelevancyMetric(threshold=t["answer_relevancy"])
    groundedness = FaithfulnessMetric(threshold=t["groundedness"])
    completeness = GEval(
        name="Completeness",
        criteria="Does the response cover every element required by the Jira acceptance criteria?",
        evaluation_params=["input", "actual_output", "expected_output"],
        threshold=t["completeness"],
    )
    correctness = GEval(
        name="Correctness",
        criteria="Is the response factually and procedurally correct against the golden answer?",
        evaluation_params=["input", "actual_output", "expected_output"],
        threshold=t["correctness"],
    )
    tool_sequence = ToolSequenceAccuracy(threshold=t["tool_sequence_accuracy"])

    results = []
    for metric in (
        (answer_relevancy, False),
        (groundedness, True),
        (completeness, False),
        (correctness, case.high_risk),
        (tool_sequence, True),
    ):
        m, hard = metric
        m.measure(tc)
        results.append(
            MetricResult(
                metric=canonical_metric_name(m.__class__.__name__.lower()),
                score=float(m.score),
                threshold=float(m.threshold),
                hard_gate=hard,
                passed=bool(m.is_successful()),
            )
        )
    return results


def _stub_results(case: EvalInput, thresholds: dict[str, float] | None) -> list[MetricResult]:
    """Deterministic stub for local dev without deepeval installed."""
    t = {**DEFAULT_THRESHOLDS, **(thresholds or {})}
    out: list[MetricResult] = []
    for name, hard in (
        ("answer_relevancy", False),
        ("groundedness", True),
        ("completeness", False),
        ("correctness", case.high_risk),
        ("tool_sequence_accuracy", True),
    ):
        score = 1.0 if case.tools_called == case.expected_tools else 0.6
        threshold = t.get(name, DEFAULT_THRESHOLDS[name])
        out.append(
            MetricResult(
                metric=name,
                score=score,
                threshold=threshold,
                hard_gate=hard,
                passed=score >= threshold,
            )
        )
    return out


class ToolSequenceAccuracy(GEval):
    """Custom metric: tool-sequence correctness (spec §7)."""

    def __init__(self, threshold: float = 0.9):
        super().__init__(
            name="ToolSequenceAccuracy",
            criteria=(
                "1) All required tools from expected_tools were called. "
                "2) No destructive/irreversible tool was called out of expected order. "
                "3) No unnecessary tool calls occurred."
            ),
            evaluation_params=["tools_called", "expected_tools"],
            threshold=threshold,
        )

File: ai-qa-automation-platform/apps/web/lib/eval/metrics.ts
import { completeJson } from "../llm/json";
import { getProvider } from "../llm/client";

/**
 * Evaluation — LLM-judge metrics (DeepEval-style) that run in serverless.
 * DeepEval itself requires Python + local execution, so these mirror the
 * five spec metrics as LLM judge calls. Deterministic fallback when the
 * LLM is unavailable: tool-sequence comparison (pass/fail).
 */
export type MetricName =
  | "answer_relevancy"
  | "groundedness"
  | "completeness"
  | "correctness"
  | "tool_sequence_accuracy";

export const DEFAULT_THRESHOLDS: Record<MetricName, number> = {
  answer_relevancy: 0.8,
  groundedness: 0.9,
  completeness: 0.75,
  correctness: 0.85,
  tool_sequence_accuracy: 0.9,
};

export const HARD_GATE_METRICS: MetricName[] = ["groundedness", "correctness", "tool_sequence_accuracy"];

export interface MetricResult {
  metric: MetricName;
  score: number;
  threshold: number;
  hardGate: boolean;
  passed: boolean;
}

export interface EvalInput {
  input: string;
  actualOutput: string;
  retrievalContext: string[];
  toolsCalled: string[];
  expectedTools: string[];
  expectedOutput?: string;
  highRisk?: boolean;
}

const CRITERIA: Record<MetricName, string> = {
  answer_relevancy: "Rate how relevant the output is to the input, 0-1.",
  groundedness: "Rate how well the output is grounded in the retrieval context, 0-1.",
  completeness: "Does the output cover every element of the expected output? Rate 0-1.",
  correctness: "Is the output factually and procedurally correct against the expected output? Rate 0-1.",
  tool_sequence_accuracy:
    "1) All required tools were called. 2) No destructive tool out of order. 3) No unnecessary calls. Rate 0-1.",
};

export async function evaluateCase(
  input: EvalInput,
  thresholds: Partial<Record<MetricName, number>> = {},
): Promise<MetricResult[]> {
  const t = { ...DEFAULT_THRESHOLDS, ...thresholds };
  const metrics: MetricName[] = [
    "answer_relevancy",
    "groundedness",
    "completeness",
    "correctness",
    "tool_sequence_accuracy",
  ];

  const results: MetricResult[] = [];
  for (const metric of metrics) {
    const isHard = HARD_GATE_METRICS.includes(metric);
    // Correctness is hard only on high-risk flows (mirrors spec).
    const effectiveHard = metric === "correctness" ? !!input.highRisk : isHard;
    results.push(await judgeMetric(input, metric, t[metric], effectiveHard));
  }
  return results;
}

async function judgeMetric(
  input: EvalInput,
  metric: MetricName,
  threshold: number,
  hardGate: boolean,
): Promise<MetricResult> {
  // Deterministic fallback when no LLM configured.
  if (!process.env.LLM_API_KEY) {
    const score = input.toolsCalled.join(",") === input.expectedTools.join(",") ? 1 : 0.6;
    return { metric, score, threshold, hardGate, passed: score >= threshold };
  }

  try {
    const provider = getProvider();
    const prompt = [
      `Metric: ${metric}.`,
      CRITERIA[metric],
      `input: ${input.input}`,
      `actual_output: ${input.actualOutput.slice(0, 4000)}`,
      input.retrievalContext.length ? `retrieval_context: ${input.retrievalContext.join("\n").slice(0, 4000)}` : "",
      input.expectedOutput ? `expected_output: ${input.expectedOutput.slice(0, 4000)}` : "",
      `tools_called: ${input.toolsCalled.join(", ")}`,
      `expected_tools: ${input.expectedTools.join(", ")}`,
      'Respond with JSON: {"score": number between 0 and 1, "reason": str}',
    ]
      .filter(Boolean)
      .join("\n");
    const raw = await provider.complete({ prompt, temperature: 0, jsonMode: true });
    const parsed = JSON.parse(raw.slice(raw.indexOf("{"), raw.lastIndexOf("}") + 1));
    const score = Math.max(0, Math.min(1, Number(parsed.score ?? 0)));
    return { metric, score, threshold, hardGate, passed: score >= threshold };
  } catch {
    // Fail-closed: LLM unavailable → deterministic tool-sequence check.
    const score = input.toolsCalled.join(",") === input.expectedTools.join(",") ? 1 : 0.6;
    return { metric, score, threshold, hardGate, passed: score >= threshold };
  }
}
